### PR TITLE
Azure HTTP Functions compliant with TCK 

### PR DESCRIPTION
**Pull request by timyates**
This PR switches the Azure function handlers to use servlet-core in much the same way as the AWS module does.

Spotted working on this that in a previous PR https://github.com/micronaut-projects/micronaut-azure/pull/398 we broke the test code that we generate in Micronaut Starter (see https://github.com/micronaut-projects/micronaut-azure/pull/509#issuecomment-1601058441)

I changed the tests in this module here https://github.com/micronaut-projects/micronaut-azure/pull/509/files#diff-5477701f4559e28df2025cfa073a4ca34eb84685ed452c0567f5c677f040a602

But I am not sure if it's correct...

I have tested this running a test project on Azure, and I *think* it's right, but it could do with someone who is used to using Azure to check we haven't broken anything important here

I also pulled in the context-path fix from #493 so we don't miss it (or someone else doesn't have to struggle with the merging)

Looking into `DefaultExecutionContext` to see why using it fails tests in third party apps

### CODE REVIEW

**Comment by timyates** [comment]
Right, this doesn't work as starter generates code such as:

```
        HttpRequestMessageBuilder.AzureHttpResponseMessage response =
            function.request(HttpMethod.GET, "/demo")
                    .invoke()
```

This was broken by https://github.com/micronaut-projects/micronaut-azure/pull/398

**Comment by timyates** [comment]
So I think it's an issue with getNativeResponse 🤔

**Comment by timyates** [comment]
Raised https://github.com/micronaut-projects/micronaut-starter/issues/1890 to deal with starter generating test code that doesn't work with 4.0.0

File: azure-function-http/src/main/java/io/micronaut/azure/function/http/MapListOfStringAndMapStringConvertibleMultiValue.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Implementations of {@link ConvertibleMultiValues} which takes a Map with key string and value List of strings and another map with key string and value string to support classes such as {@link com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent}.
```
Comment (wetted) [inline azure-function-http/src/main/java/io/micronaut/azure/function/http/MapListOfStringAndMapStringConvertibleMultiValue.java:35]: Is this link correct. When I checkout the branch it's not reachable.

File: test-suite-azure-function-http-context-path/src/test/resources/logback.xml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+             ch.qos.logback.classic.encoder.PatternLayoutEncoder by default -->
+        <encoder>
+            <pattern>%cyan(%d{HH:mm:ss.SSS}) %gray([%thread]) %highlight(%-5level) %magenta(%logger{36}) - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+    <logger name="io.micronaut.web.router" level="trace" />
```
Comment (wetted) [inline test-suite-azure-function-http-context-path/src/test/resources/logback.xml:15]: does `trace` logging need to remain?

**Review by wetted — APPROVED** [review]
Looks like a real nice piece of work. Congrats on clearing all the TCK tests.

File: test-suite-azure-function-http-context-path/src/test/resources/logback.xml (hunk 2 above)
Comment (timyates) [reply test-suite-azure-function-http-context-path/src/test/resources/logback.xml:15]: It's useful when it's broken as you see the paths the router has created in the logs

**Review by sdelamo — COMMENTED** [review]
We need to document you can bind: 

`com.microsoft.azure.functions.ExecutionContext`
`com.microsoft.azure.functions.HttpRequestMessage`
`com.microsoft.azure.functions.TraceContext`

File: azure-function-http/src/main/java/io/micronaut/azure/function/http/AzureCookies.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Implementation of {@link Cookies} for serverless.
+ *
+ */
+@Internal
+public final class AzureCookies implements Cookies {
```
Comment (sdelamo) [inline azure-function-http/src/main/java/io/micronaut/azure/function/http/AzureCookies.java:40]: I think we should move this class to `io.micronaut.netty.cookies` in core.

**Review by sdelamo — CHANGES_REQUESTED** [review]
It misses documentation about change.

File: azure-function-http/src/main/java/io/micronaut/azure/function/http/BinaryContentConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Bean to check if response content is binary and should be base 64 encoded.
+ */
+@Internal
+@Singleton
+public final class BinaryContentConfiguration {
```
Comment (sdelamo) [inline azure-function-http/src/main/java/io/micronaut/azure/function/http/BinaryContentConfiguration.java:29]: We have this class in AWS as well. @timyates Can we move this class to servlet-core: 

https://github.com/micronaut-projects/micronaut-aws/blob/master/function-aws-api-proxy/src/main/java/io/micronaut/function/aws/proxy/BinaryContentConfiguration.java

**Comment by sdelamo** [comment]
@graemerocher can you review this PR? It currently passes every TCK test and I have changed it to tell users they have to remove the azure api route prefix and then they can use normally `micronaut.server.context-path` if they wish for a prefix.

**Review by sdelamo — APPROVED** [review]

**Comment by sdelamo** [comment]
> We need to document you can bind:
> 
> `com.microsoft.azure.functions.ExecutionContext` `com.microsoft.azure.functions.HttpRequestMessage` `com.microsoft.azure.functions.TraceContext`

I have created an issue for this.

File: azure-function-http/src/main/java/io/micronaut/azure/function/http/AzureCookies.java (hunk 3 above)
Comment (sdelamo) [reply azure-function-http/src/main/java/io/micronaut/azure/function/http/AzureCookies.java:40]: @timyates create an issue for this and add it to [`5.1.0 Release`  project](https://github.com/orgs/micronaut-projects/projects/53/views/1).

File: azure-function-http/src/main/java/io/micronaut/azure/function/http/BinaryContentConfiguration.java (hunk 4 above)
Comment (sdelamo) [reply azure-function-http/src/main/java/io/micronaut/azure/function/http/BinaryContentConfiguration.java:29]: @timyates create an issue for this and add it to [5.1.0 Release project](https://github.com/orgs/micronaut-projects/projects/53/views/1).

Comment (timyates) [reply azure-function-http/src/main/java/io/micronaut/azure/function/http/BinaryContentConfiguration.java:29]: https://github.com/micronaut-projects/micronaut-servlet/pull/504

File: azure-function-http/src/main/java/io/micronaut/azure/function/http/HttpResponseMessageAdapter.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @Override
+    public Optional<T> getBody() {
+        return (Optional<T>) Optional.ofNullable(message.getBody());
+    }
+
+    @Override
+    public <B> MutableHttpResponse<B> body(B body) {
+        return (MutableHttpResponse<B>) this;
```
[the remark this answers is not in the document]
Comment (timyates) [reply azure-function-http/src/main/java/io/micronaut/azure/function/http/HttpResponseMessageAdapter.java]: We do the same in AWS.  This class is only used in the TCK tests...  Maybe we should move it out to the TCK

Done that in https://github.com/micronaut-projects/micronaut-azure/pull/509/commits/04ad2a19871c1555e72f4e2f391f6f955b9ae47a

**Review by graemerocher — APPROVED** [review]

**Pull request merged**